HRMP Channel Setup Documentation (#34)
papermoonio/polkadot-ecosystem-docs-draft: docs/build-on-polkadot/xcm/hrmp-channels/overview.md

---
title: HRMP Channels

> description: HRMP channels enable cross-chain communication in Polkadot, a temporary solution before the more efficient XCMP protocol is implemented.

Reviewer: It seems that `index.md` description and `overview.md` description are the same, I think this would impact SEO

---

# HRMP Channels

## Introduction

Polkadot is designed to enable seamless interoperability between its connected parachains. At the core of this interoperability is the Cross-Consensus Message Format (XCM), a standard language that allows parachains to communicate and interact with each other.

> The actual network-layer protocol responsible for delivering XCM-formatted messages between parachains is called the Cross-Chain Message Passing (XCMP) protocol. XCMP maintains messaging queues on the relay chain, serving as a bridge to facilitate cross-chain interactions.

Reviewer: ```suggestion
The network-layer protocol responsible for delivering XCM-formatted messages between parachains is the Cross-Chain Message Passing (XCMP) protocol. XCMP maintains messaging queues on the relay chain, serving as a bridge to facilitate cross-chain interactions.
```

However, as XCMP is still under development, Polkadot has implemented a temporary substitute known as Horizontal Relay-routed Message Passing (HRMP). HRMP provides the same interface and functionality as the planned XCMP, but with a key difference - it stores all messages directly in the relay chain's storage, which is more resource-intensive.

Once XCMP is fully implemented, HRMP will be deprecated in favor of the native XCMP protocol. XCMP will offer a more efficient and scalable solution for cross-chain message passing, as it will not require the relay chain to store all the messages.

## Prerequisites (WIP)

<!-- TODO -->

## Establishing HRMP Channels

To facilitate communication between parachains using the HRMP protocol, the parachains must explicitly establish communication channels by registering them on the relay chain.

Downward and upward channels from and to the relay chain are implicitly available, meaning they do not need to be explicitly opened.

> Opening an HRMP channel requires the parachains involved to make a deposit on the relay chain. This deposit covers the expenses of using the relay chain's storage for the message queues associated with the channel.

Reviewer: Is this deposit a fixed amount, or may it vary depending on the network, the messages to be sent, and so on?

### Relay Chain Parameters

> Each Polkadot relay chain has a set of configurable parameters that control the behavior of the message channels between parachains. These parameters include `hrmpSenderDeposit`, `hrmpRecipientDeposit`, `hrmpChannelMaxMessageSize`, `hrmpChannelMaxCapacity` and more.

Reviewer: ```suggestion
Each Polkadot relay chain has a set of configurable parameters that control the behavior of the message channels between parachains. These parameters include `hrmpSenderDeposit`, `hrmpRecipientDeposit`, `hrmpChannelMaxMessageSize`, `hrmpChannelMaxCapacity`, and more.
```

> When a parachain wants to open a new channel, it must take these parameter values into account to ensure the channel is configured properly.

Reviewer: ```suggestion
When a parachain wants to open a new channel, it must consider these parameter values to ensure the channel is configured correctly.
```

To view the current values of these parameters in the Polkadot network:

> 1. Visit [Polkadot.js Apps](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fpolkadot.api.onfinality.io%2Fpublic-ws#/explorer), navigate to the **Developer** dropdown and select the **Chain state** option

Reviewer: In one of the tutorials I wrote, I've been requested to break down this into more granular actions. Maybe something like the following would work: 

```
1. Query Chain State
      a. Navigate to the **Developer** by using [Polkadot.js Apps](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fpolkadot.api.onfinality.io%2Fpublic-ws#/explorer)
      b. Select the **Chain state** option
```

However, this implies an update to the image to include the steps corresponding to those actions.

    ![](/polkadot-ecosystem-docs-draft/images/build-on-polkadot/hrmp-channels/hrmp-channels-1.webp)

> 2. Under the **configuration** section, call the `activeConfig()` function by clicking the button with the **+** icon. This will display the current settings for all the Polkadot network parameters, including the HRMP channel settings

Reviewer: Same as above

    ![](/polkadot-ecosystem-docs-draft/images/build-on-polkadot/hrmp-channels/hrmp-channels-2.webp)

> ### Dispatching Extrinsics

Reviewer 2: Here it is not correct right? Beause we use camel case for title?

> ### Dispatching Extrinsics

Author: Yep, this should be fixed

Establishing new HRMP channels between parachains requires dispatching specific extrinsic calls on the Polkadot relay chain from the parachain's origin.

The most straightforward approach is to implement the channel opening logic off-chain, then use the XCM pallet's `send` extrinsic to submit the necessary instructions to the relay chain. However, the ability to send arbitrary programs through the `Transact` instruction in XCM is typically restricted to privileged origins, such as the sudo pallet or governance mechanisms.

Parachain developers have a few options for triggering the required extrinsic calls from their parachain's origin, depending on the configuration and access controls defined:

- Sudo - if the parachain has a sudo pallet configured, the sudo key holder can use the sudo extrinsic to dispatch the necessary channel opening calls
- Governance - the parachain's governance system, such as a council or OpenGov, can be used to authorize the channel opening calls
- Privileged Accounts - the parachain may have other designated privileged accounts that are allowed to dispatch the HRMP channel opening extrinsics

## Opening HRMP Channels Between Parachains

> For establishing communication channels between parachains on the Polkadot network, using the HRMP protocol, the following steps are required:

Reviewer: ```suggestion
For establishing communication channels between parachains on the Polkadot network using the HRMP protocol, the following steps are required:
```

1. Channel request - the parachain that wants to open an HRMP channel must make a request to the parachain it wishes to have an open channel with
2. Channel acceptance - the other parachain must then accept this request in order to complete the channel establishment

This process results in a unidirectional HRMP channel, where messages can flow in only one direction between the two parachains.

> To enable bidirectional communication, an additional HRMP channel must be established in the opposite direction. This requires repeating the request and acceptance process, but with the parachains reversing their roles.

Reviewer: ```suggestion
An additional HRMP channel must be established in the opposite direction to enable bidirectional communication. This requires repeating the request and acceptance process but with the parachains reversing their roles.
```

> Once both unidirectional channels are established, the parachains can then send messages back and forth freely through the bidirectional HRMP communication channel.

Reviewer: ```suggestion
Once both unidirectional channels are established, the parachains can send messages back and forth freely through the bidirectional HRMP communication channel.
```

### Procedure for Initiating HRMP Channel Setup

> This example will demonstrate how to open a channel between parachain 2500 and parachain 2600, using Rococo Local as the relay chain.

Reviewer: Maybe to make the tutorial even more generic, you can say something like: 

`between the dispatcher (parachain 2500 in this tutorial) and the receiver (parachain 2600)`

For the rest of the tutorial, we can refer to them as dispatcher and receiver directly

For sure, dispatcher and receiver are the first words that came into my mind, but you understand my point.

#### Step 1 - Fund Sovereign Account {: #init-fund-sovereign-account }
<!-- <a id="init-fund-sovereign-account"></a> -->

> The sovereign account for parachain 2500 on the relay chain must be funded so it can take care of any XCM transact fees.

Reviewer: I would include a brief explanation of what a sovereign account is, or provide a reference to an external source that explains it.

Use Polkadot.js Apps UI to connect to the relay chain and transfer funds from your account to the parachain 2500 sovereign account.

> ![](/polkadot-ecosystem-docs-draft/images/build-on-polkadot/hrmp-channels/hrmp-channels-3.webp)

Reviewer: I think an arrow is missing in this image, maybe an arrow that point to **Make Transfer**

    3. Fill in the parameters
        - **recipient** - parachain ID of the target chain (in this case, 2600)
        - **proposedMaxCapacity** - max number of messages that can be pending in the channel at once
        - **proposedMaxMessageSize** - max message size that could be put into the channel
    4. Copy the encoded call data
    ![](/polkadot-ecosystem-docs-draft/images/build-on-polkadot/hrmp-channels/hrmp-channels-5.webp)
    The encoded call data for opening a channel with parachain 2600 is `0x3c00280a00000800000000001000`.

#### Step 3 - Crafting and Submitting the XCM Message {: #init-crafting-and-submitting-the-xcm-message }

To initiate the HRMP channel opening process, you need to create an XCM message that includes the encoded `hrmpInitOpenChannel` call data from the previous step. This message will be sent from your parachain to the relay chain.

This example uses the `sudo` pallet to dispatch the extrinsic. Verify the XCM configuration of the parachain you're working with and ensure you're using an origin with the necessary privileges to execute the `polkadotXcm.send` extrinsic.

The XCM message should contain the following instructions:

> - **WithdrawAsset** - withdraws assets from the origin's ownership and places them in the Holding Register

Reviewer: I'd add a reference to some link that explains what is the Holding Register

- **BuyExecution** - pays for the execution of the current message using the assets in the Holding Register
- **Transact** - execute the encoded transaction call
- **RefundSurplus** - increases the Refunded Weight Register to the value of the Surplus Weight Register, attempting to reclaim any excess fees paid via BuyExecution
- **DepositAsset** - subtracts assets from the Holding Register and deposits equivalent on-chain assets under the specified beneficiary's ownership

In essence, this process withdraws funds from the parachain's sovereign account to the XCVM Holding Register, then uses these funds to purchase execution time for the XCM Transact instruction, executes Transact, refunds any unused execution time and deposits any remaining funds into a specified account.

To submit the XCM message:

1. Connect to parachain 2500
   
2. Construct the XCM message by filling in all required parameters. Make sure to replace the **call** field with your encoded `hrmpInitOpenChannel` call data from the previous step
   
3. Click the **Submit Transaction** button to dispatch the XCM message to the relay chain.

> ![](/polkadot-ecosystem-docs-draft/images/build-on-polkadot/hrmp-channels/hrmp-channels-6.webp)

Reviewer: It's weird that this image has direct steps 2 and 3. I understand that it's mapping the sentences above, but maybe to make it even more precise, you can do the following:

```
To submit the XCM message, ensure you are connected to parachain 2500. Then: 

1. Construct the XCM message by filling in all required parameters. Make sure to replace the **call** field with your encoded `hrmpInitOpenChannel` call data from the previous step

2. Click the **Submit Transaction** button to dispatch the XCM message to the relay chain.
```

!!! note
    The exact process and parameters for submitting this XCM message may vary depending on your specific parachain and relay chain configurations. Always refer to the most current documentation for your particular network setup.

After submitting the XCM message to initiate the HRMP channel opening, you should verify that the request was successful. Follow these steps to check the status of your channel request:

1. Using Polkadot.js Apps, connect to the relay chain and navigate to the **Developer** dropdown, then select the **Chain state** option

> 2. Under the **hrmp** section, select the `hrmpOpenChannelRequests` function. Click the button with the **+** icon next to it to call the function. This will display the status of all pending channel requests

Reviewer: break this down

If your channel request was successful, you should see an entry for your parachain ID in the list of open channel requests. This confirms that your request has been properly registered on the relay chain and is awaiting acceptance by the target parachain

> ![](/polkadot-ecosystem-docs-draft/images/build-on-polkadot/hrmp-channels/hrmp-channels-7.webp)

Reviewer: I'd move this image above the explanation

    4. Copy the encoded call data
    ![](/polkadot-ecosystem-docs-draft/images/build-on-polkadot/hrmp-channels/hrmp-channels-8.webp)
    The encoded call data for accepting a channel with parachain 2500 should be `0x3c01c4090000`

#### Step 3 - Crafting and Submitting the XCM Message

To accept the HRMP channel opening, you need to create and submit an XCM message that includes the encoded `hrmpAcceptOpenChannel` call data from the previous step. This process is similar to the one described in the previous section's [Step 3 - Crafting and Submitting the XCM Message](#init-crafting-and-submitting-the-xcm-message), with a few key differences:
- Use the encoded call data for `hrmpAcceptOpenChannel` obtained in Step 2 of this section
- In the last XCM instruction (DepositAsset), set the beneficiary to parachain 2600's sovereign account to receive any surplus funds

To submit the XCM message:

1. Connect to parachain 2600 using Polkadot.js Apps
2. Construct the XCM message by filling in all required parameters, ensuring you use the correct encoded call data and beneficiary information
3. Submit the transaction to dispatch the XCM message to the relay chain

>     ![](/polkadot-ecosystem-docs-draft/images/build-on-polkadot/hrmp-channels/hrmp-channels-9.webp)

Reviewer: This image should have steps instead of arrows
